pkg/storegateway/sharding_strategy.go

```diff
@@ -115,30 +117,49 @@ func (s *ShuffleShardingStrategy) FilterUsers(_ context.Context, userIDs []strin
 }
 
 // FilterBlocks implements ShardingStrategy.
-func (s *ShuffleShardingStrategy) FilterBlocks(_ context.Context, userID string, metas map[ulid.ULID]*metadata.Meta, synced *extprom.TxGaugeVec) error {
+func (s *ShuffleShardingStrategy) FilterBlocks(_ context.Context, userID string, metas map[ulid.ULID]*metadata.Meta, loaded map[ulid.ULID]struct{}, synced *extprom.TxGaugeVec) error {
 	subRing := GetShuffleShardingSubring(s.r, userID, s.limits)
-	filterBlocksByRingSharding(subRing, s.instanceAddr, metas, synced, s.logger)
+	filterBlocksByRingSharding(subRing, s.instanceAddr, metas, loaded, synced, s.logger)
 	return nil
 }
 
-func filterBlocksByRingSharding(r ring.ReadRing, instanceAddr string, metas map[ulid.ULID]*metadata.Meta, synced *extprom.TxGaugeVec, logger log.Logger) {
+func filterBlocksByRingSharding(r ring.ReadRing, instanceAddr string, metas map[ulid.ULID]*metadata.Meta, loaded map[ulid.ULID]struct{}, synced *extprom.TxGaugeVec, logger log.Logger) {
 	bufDescs, bufHosts, bufZones := ring.MakeBuffersForGet()
 
 	for blockID := range metas {
 		key := cortex_tsdb.HashBlockID(blockID)
-		set, err := r.Get(key, BlocksSync, bufDescs, bufHosts, bufZones)
-
-		// If there are no healthy instances in the replication set or
-		// the replication set for this block doesn't include this instance
-		// then we filter it out.
-		if err != nil || !set.Includes(instanceAddr) {
-			if err != nil {
-				level.Warn(logger).Log("msg", "excluded block because failed to get replication set", "block", blockID.String(), "err", err)
-			}
 
+		// A block needs to be always loaded if is owned by the store-gateway itself.
+		set, err := r.Get(key, BlocksOwnerSync, bufDescs, bufHosts, bufZones)
+		if err != nil {
+			level.Warn(logger).Log("msg", "excluded block because failed to get replication set", "block", blockID.String(), "err", err)
```

Comment:
Now that I think about this, a better approach may be keep it loaded if it was previously loaded. That would be an improvement over `master` version.